Add ClusterClass Support for Managed Clusters (#4155)
kubernetes-sigs/cluster-api-provider-azure: docs/book/src/topics/clusterclass.md

# ClusterClass

> - **Feature status:** GA

Reviewer: ClusterClass isn't GA, it requires a feature gate?

Author: oops!

- **Feature gate:** `ClusterTopology=true`

> [ClusterClass](https://cluster-api.sigs.k8s.io/tasks/experimental-features/cluster-class/index.html) is a collection of templates that define a topology (control plane and machine deployments) to be used to continuously reconcile one or more Clusters. It is a new Cluster API feature that is built on top of the existing Cluster API resources and provides a set of tools and operations to streamline cluster lifecycle management while maintaining the same underlying API.

Reviewer: ```suggestion
[ClusterClass](https://cluster-api.sigs.k8s.io/tasks/experimental-features/cluster-class/index.html) is a collection of templates that define a topology (control plane and machine deployments) to be used to continuously reconcile one or more Clusters. It is built on top of the existing Cluster API resources and provides a set of tools and operations to streamline cluster lifecycle management while maintaining the same underlying API.
```

the new part will get old quickly

# ClusterClass

- **Feature status:** GA

> - **Feature gate:** MachinePool=true ClusterTopology=true

Reviewer 2: ```suggestion
- **Feature gate:** `MachinePool=true`, `ClusterTopology=true`
```

> - **Feature gate:** `MachinePool=true`, `ClusterTopology=true`

Reviewer 3: MachinePool isn't required for all cluster class usage right? Just if you need machine pools?

CAPZ currently supports ClusterClass for both managed (AKS) and self-managed clusters. CAPZ implements this with three custom resources:
1. AzureClusterTemplate
2. AzureManagedClusterTemplate
3. AzureManagedControlPlaneTemplate

Each resource is a template for the corresponding CAPZ resource. For example, the AzureClusterTemplate is a template for the CAPZ AzureCluster resource. The template contains a set of parameters that are able to be shared across multiple clusters.

## Deploying a Self-Managed Cluster with ClusterClass

To deploy a self-managed cluster with ClusterClass, you must first create a ClusterClass resource. The ClusterClass resource defines the cluster topology, including the control plane and machine deployment templates. The ClusterClass resource also defines the parameters that can be used to customize the cluster topology. 

Please refer to the Cluster API book for more information on how to write a ClusterClass topology: https://cluster-api.sigs.k8s.io/tasks/experimental-features/cluster-class/write-clusterclass.html

> For a self-managed cluster, the AzureClusterTemplate is used to define the Azure infrastructure for the cluster. The following example shows a basic AzureClusterTemplate resource:

Reviewer 2: ```suggestion
For a self-managed cluster, the AzureClusterTemplate defines the Azure infrastructure for the cluster. The following example shows a basic AzureClusterTemplate resource:
```

The following example shows a basic AzureManagedClusterTemplate and AzureManagedControlPlaneTemplate resource:

```yaml
apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
kind: AzureManagedClusterTemplate
metadata:
  name: capz-clusterclass-cluster
---
apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
kind: AzureManagedControlPlaneTemplate
metadata:
  name: capz-clusterclass-control-plane
spec:
  location: westus2
  subscriptionID: 00000000-0000-0000-0000-000000000000
  version: 1.25.2
```

## Excluded Fields

Since a ClusterClass is a template for a Cluster, there are some fields that are not allowed to be shared across multiple clusters. For each of the ClusterClass resources, the following fields are excluded:

### AzureClusterTemplate

> - `spec.resourceGroup`

Reviewer 3: Is there a better place we can refer to these fields? Or do we need to document them explicitly at all? Listing them in the docs like this is sure to go stale, especially has we continue to add features.

Author: I think we can just leave it out, since users can refer to the CRDs for available fields
